src/main/java/com/ning/http/client/providers/grizzly/ConnectionManager.java

```diff
@@ -214,16 +219,38 @@ boolean isReadyInPool(final Connection c) {
         final ConnectionInfo<SocketAddress> ci = pool.getConnectionInfo(c);
         return ci != null && ci.isReady();
     }
-    
+
     static boolean isKeepAlive(final Connection connection) {
         return !IS_NOT_KEEP_ALIVE.isSet(connection);
     }
-    
+
     private static String getPartitionId(InetAddress overrideAddress, Request request,
             ProxyServer proxyServer) {
-        return (overrideAddress != null ? overrideAddress.toString() + "_" : "") + 
-            request.getConnectionPoolPartitioning()
-             .getPartitionKey(request.getUri(), proxyServer).toString();
+        String partitionId = (overrideAddress != null ? overrideAddress.toString() + "_" : "") +
+                   request.getConnectionPoolPartitioning()
+                           .getPartitionKey(request.getUri(), proxyServer).toString();
+
+        // In the NTLM case, add in the partitionId the authentication credentials, since one
+        // connection per target-credentials pair should be used.
+        if (request.getRealm() != null &&
+            request.getRealm().getScheme().equals(Realm.AuthScheme.NTLM)) {
+            Realm requestRealm = request.getRealm();
+            partitionId = partitionId
+                    .concat("_")
+                    .concat(requestRealm.getPrincipal())
+                    .concat(requestRealm.getPassword())
+                    .concat(requestRealm.getNtlmDomain())
+                    .concat(requestRealm.getNtlmHost());
```

Comment:
Maybe I should use some kind of hash of all credentials, or encode them in some way. What do you think @afelisatti ?

Reply:
Is the password needed for the realmId. Isn't the principal and the domain enough?

Reply:
Ok, in case a change in the password to an incorrect passsword is provided. A salted secure hashing would do I think.

Reply:
After discussion, an unsalted hash should be used.